Add readiness tracker to ensure caches have been loaded before serving traffic

shomron opened the pull request:
This commit adds readiness.Tracker, responsible for managing
expectations and observations of templates, constraints, and data.
Expectations are set according to the initial state of the cluster when
Gatekeeper starts - pre-existing templates and their corresponding
constraints are tracked until they are loaded by their respective
controllers into OPA. Similarly, cached data resources specified by
the Config singleton are also tracked to completion.

Templates or configured resource removed during warm-up are removed from
the initial expected set and will not block readiness.

The tracker is registered as a readiness probe - traffic will not be
routed to the admission controller while the readiness probe is not
returning success.

Fixes: #405
Signed-off-by: Oren Shomron <shomron@gmail.com>

shomron commented:
@maxsmythe @sozercan @ritazh pushing this mostly-complete WIP in case you want an early review. I need to fix an outstanding issue where status can go from ready->unready when new constraints are added after the fact. I hope to get that out shortly.

maxsmythe commented on pkg/controller/config/config_controller.go:
```diff
@@ -18,9 +18,13 @@ package config
+	"sync"
 	"time"
 
+	"github.com/open-policy-agent/gatekeeper/pkg/readiness"
```
nit: import  consolidation

maxsmythe commented on pkg/controller/constraint/constraint_controller.go:
```diff
@@ -298,11 +308,22 @@ func logRemoval(l logr.Logger, constraint *unstructured.Unstructured, enforcemen
 	unstructured.RemoveNestedField(obj.Object, "status")
 	_, err := r.opa.AddConstraint(context.Background(), obj)
-	return err
+	if err != nil {
```
Are there some errors that should not lead to expectations being canceled? The top candidate in my mind is when there is a failure due to a missing template b/c the template has not yet been added to OPA.

That being said, I think the way we have the timing, the template needs to be added before the watch starts. I'm unsure if relying on that internal in a separate module is too brittle.

shomron replied:
As discussed, let's consider adding configuration knobs to tune this behavior as a follow-up.

maxsmythe commented on pkg/controller/constrainttemplate/constrainttemplate_controller.go:
```diff
@@ -341,13 +356,19 @@ func (r *ReconcileConstraintTemplate) handleUpdate(
 			log.Error(err, "failed to report constraint template ingestion duration")
 		}
 		err := r.reportErrorOnCTStatus("ingest_error", "Could not ingest Rego", ct, err)
+		r.tracker.CancelTemplate(unversionedCT) // Don't track templates that failed compilation
```
I'm wondering if we should have an error count threshold here. We don't want to hard-down the webhook on a bad template, but we also don't want to relax the warming requirement due to transient errors. Keeping a counter inside of tracker that expires a resource after a certain threshold seems like it may strike the right balance.

shomron replied:
Let's address as a followup - it's hard to differentiate between transient and non-transient errors, let's continue to explore what the right balance will be here.

ritazh replied:
Can you pls create a followup issue for this?

shomron replied:
https://github.com/kubernetes-sigs/controller-runtime/issues/948

maxsmythe commented on pkg/controller/constrainttemplate/constrainttemplate_controller.go:
```diff
@@ -341,13 +356,19 @@ func (r *ReconcileConstraintTemplate) handleUpdate(
 
+	// Mark for readiness tracking
+	t := r.tracker.For(gvkConstraintTemplate)
+	t.Observe(unversionedCT)
```
Is there something like `setDataExpectations` for constraints and templates as well? After a quick look I don't see any roster gathering code.

shomron replied:
Yes, it's in [`trackConstraintTemplates`](https://github.com/open-policy-agent/gatekeeper/blob/a22c7e5638745b22d0ef9ea31e00a08d47f150ba/pkg/readiness/ready_tracker.go#L218).

maxsmythe replied:
Thanks! Is there a race condition putting it in Run(), given that other controllers and the server (who is executing this health check) are also starting up in parallel?

shomron replied:
`CancelTemplates` is synchronized with `Run` via the `t.ready` channel because it depends on `constraintTemplates` being initialized. Other than that, objectTrackers are internally synchronized. Observations are allowed to occur before expectations. There is a matter of avoiding new constraint expectations once the original roster is locked in - I'm thinking about different approaches to address this now.

shomron replied:
The above is resolved now.

maxsmythe commented on pkg/controller/config/config_controller.go:
```diff
@@ -264,6 +288,43 @@ func (r *ReconcileConfig) replayData(ctx context.Context, w *watch.Set) error {
+			Kind:    gvk.Kind + "List",
+		})
+		err := r.reader.List(ctx, u)
+		if err != nil {
```
How do we avoid transient errors leading to skipping cache warming here?

shomron replied:
This is similar to your question above - how many errors is too many errors, how do we distinguish between a recoverable and an unrecoverable error, and how long do we want to hold up the controller in pursuit of the perfect.

maxsmythe replied:
IMO the answer is more than one, as this is a tradeoff between availability and security and 1 is about as hard toward availability as we can go. Since we aim to improve security, a more conservative choice seems appropriate. The exact number is likely a judgement call.

The alternative is to discount error types we "know" are transient, but that might lead to the controller being perma-down if our assessment is inaccurate. However, one note in favor of the perma-down approach... if we assume that all non-recoverable errors are user-fixable (e.g. bad constraint template => delete that template), that may be a tradeoff some people are willing to take. I likely wouldn't choose it as a default, personally. 

The dynamics of how long it keeps the controller down... what are the levers for controlling this? Do we have control over the retry wait periods? Assuming we use retries, can we make the tolerances user tune-able?

shomron commented:
@maxsmythe @ritazh @sozercan this is ready for another look.

I've added a bit more documentation on scenarios that currently fail-open.
Regarding the watch.Manager changes to support listing, I'm not in love with these changes, and as you pointed out the lifecycle of the registrar is not sufficient to actually ensure resources are cached once. I think we could achieve that, but I might be more include to roll that part back and bypass the cache for warm-up to reduce complexity. Let me know what you think on this.

Also for future improvements we can consider releasing memory after the readiness circuit-breaker trips.

Thanks for all your time talking through these changes!

maxsmythe commented on pkg/watch/list.go:
```diff
@@ -0,0 +1,101 @@
+	"strings"
+	"time"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/syncutil"
```
nit: consolidate imports

shomron replied:
This file has been removed.

maxsmythe commented on pkg/watch/list.go:
```diff
@@ -0,0 +1,101 @@
+// List lists resources of a particular kind using a watch Manager (informers are shared with dynamic watchers).
+func List(ctx context.Context, wm *Manager, gvk schema.GroupVersionKind, cbForEach func(runtime.Object)) error {
+	initialPopulation := make(chan event.GenericEvent, 1024)
+	name := "list-" + gvk.Kind + randSuffix(8)
```
I agree this adds extra complexity without much benefit. If we want to try to do something like this, we should put the registrar on the constraint template tracker and config tracker, which would guarantee that its lifespan interacts with the dynamic sync registrars. It would also avoid creating random names.

Finally, we should be fine not to share the same cache as watchmanager, as this appears to be adding complexity to an already complex bit of code. I think I'd be more comfortable if we didn't touch watch manager TBH.

maxsmythe commented on pkg/watch/list.go:
```diff
@@ -0,0 +1,101 @@
+}
+
+// List lists resources of a particular kind using a watch Manager (informers are shared with dynamic watchers).
+func List(ctx context.Context, wm *Manager, gvk schema.GroupVersionKind, cbForEach func(runtime.Object)) error {
```
Is this `List()` or `ListFor()` (or some better name)?

maxsmythe commented on pkg/watch/manager.go:
```diff
@@ -177,6 +186,11 @@ func (wm *Manager) doAddWatch(r *Registrar, gvk schema.GroupVersionKind) error {
 		// First watcher gets a fresh informer, register for events.
 		informer.AddEventHandler(wm)
+
+		// Special case for sync watchers - replay initial set only.
```
If we need special cases to make this work, my preference is to leave this out.

maxsmythe commented on pkg/watch/manager.go:
```diff
@@ -241,8 +255,10 @@ func (wm *Manager) replaceWatches(r *Registrar) error {
 	for gvk := range desired {
-		if err := wm.doAddWatch(r, gvk); err != nil {
-			errlist = append(errlist, fmt.Errorf("adding watch for %+v %w", gvk, err))
+		for r := range desired[gvk].registrars {
```
Why has this changed to iterate over all registrars when it is called from a specific registrar?

shomron replied:
`desired` spans all registrars, so this change fixes a bug where if `r1` desires something, and we call Replace for `r2`, `r1`'s desired roster bleeds into `r2`.

If you're asking whether this could changed to:

```
for gvk := range desired {
        if _, ok := desired[gvk].registrars[r]; !ok {
                continue
        }
	if err := wm.doAddWatch(r, gvk); err != nil {
		errlist = append(errlist, fmt.Errorf("adding watch for %+v %w", gvk, err))
	}
}
```

I *think* it could - I'll squint at it a bit more. Although it might be safer to leave as is, where this function will definitely reconcile to the desired state.
Let me know what you think!

Regarding the original bug, I'll add a test case that captures it.

shomron replied:
@maxsmythe I've reworked this fix as a separate issue: #620.

maxsmythe commented on pkg/watch/replay.go:
```diff
@@ -77,21 +80,26 @@ func (wm *Manager) replayEventsLoop() error {
 				continue
 			}
 
+			if req.r.events == nil && req.r.initialPopulation == nil {
```
This dual-channel-writing logic is another exceptional case that makes me think we should disentangle the readiness checker and the watch manager

shomron commented on main.go:
```diff
@@ -256,12 +256,25 @@ func startControllers(mgr ctrl.Manager, sw *watch.ControllerSwitch, setupFinishe
 		os.Exit(1)
 	}
 
+	tracker, err := readiness.SetupTracker(mgr, wm)
```
Should we remove Healthz? Do we want to retain a basic health probe? I don't think we want to cross the wires between health and readiness, they serve different purposes.

maxsmythe replied:
No, healthz should stay, sorry. I was conflating the two. Readiness should go.

shomron commented on pkg/controller/constrainttemplate/constrainttemplate_controller_suite_test.go:
```diff
@@ -34,20 +36,29 @@ import (
 func TestMain(m *testing.M) {
+	var err error
+
+	// Load framework CRDs separately - controller-runtime tries to parse any *.yaml it sees,
```
Not sure I follow you - what is the extra loading? There are two paths loading CRDs here: `envtest.readCRDFiles` picks up all yaml files in paths specified in `CRDDirectoryPaths`, in this case the `Config` CRD.

`util.ReadCRD` which I added reads individual files, in this case the `ConstraintTemplate` CRD, and avoids reading the `kustomization.yaml` in that same directory.

Finally, `envtest.InstallCRDs` applies and waits for the resources read in the above two.

What are you looking for - are you saying you're like to remove `CRDDirectoryPaths` and explicitly load everything with `util.ReadCRD`? Or is the current state reasonable until the interface of `envtest` can be changed to allow more explicit whitelist/blacklist?

maxsmythe replied:
As you said, `CRDDirectoryPaths` gets piped into `readCRDFiles`, which supports citing specific YAML files:

https://github.com/kubernetes-sigs/controller-runtime/blob/c2dfe1abee25da5e35f5c803c82a06db6275a239/pkg/envtest/crd.go#L302-L308

My suggestion is to remove the custom CRD loading and just use envtest.

maxsmythe replied:
In the version of controller-runtime that you upgraded from, this behavior existed but was buggy (fixed in current vendored version). That was the cause of test flakiness previously.

shomron replied:
Ah, this is much better. I didn't realize they supported both directories and files. 👍

shomron commented on pkg/readiness/integration_suite_test.go:
```diff
@@ -0,0 +1,99 @@
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
```
Done.

shomron commented on pkg/controller/constrainttemplate/constrainttemplate_controller_suite_test.go:
```diff
@@ -24,6 +24,8 @@ import (
 	"github.com/onsi/gomega"
 	"github.com/open-policy-agent/gatekeeper/api"
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+
```
Done.

shomron commented:
@maxsmythe I think I've addressed all comments in the final commit. Let me know what you think.

maxsmythe commented on pkg/readiness/ready_tracker.go:
```diff
@@ -0,0 +1,372 @@
+
+const constraintGroup = "constraints.gatekeeper.sh"
+
+var cfgKey = types.NamespacedName{Namespace: util.GetNamespace(), Name: "config"} // Config resource singleton reference
```
bump nit

shomron replied:
That would lead to an import cycle. Should I move it to a third package or leave it be?

maxsmythe replied:
I say if you touch the PR again move it to a new package, but not a blocker.

maxsmythe commented on pkg/readiness/ready_tracker.go:
```diff
@@ -0,0 +1,372 @@
+
+// ForData returns Expectations for tracking data of the requested resource kind.
+func (t *Tracker) ForData(gvk schema.GroupVersionKind) Expectations {
+	// Avoid new data trackers after templates have been satisfied.
```
nit: templates -> config, satisfied -> populated

maxsmythe commented on pkg/readiness/ready_tracker.go:
```diff
@@ -0,0 +1,372 @@
+		return t.config
+	}
+
+	// Avoid new constraint trackers after templates have been satisfied.
```
nit: satisfied -> populated

maxsmythe commented on pkg/readiness/ready_tracker.go:
```diff
@@ -0,0 +1,372 @@
+			return nil // do not return an error, this would abort other constraint trackers!
+		})
+	}
+	t.templates.ExpectationsDone()
```
bump question

shomron replied:
Changed.

maxsmythe commented on pkg/readiness/ready_tracker.go:
```diff
@@ -0,0 +1,372 @@
+		return err
+	}
+
+	constraints.ExpectationsDone()
```
bump

shomron replied:
Yes I agree.

maxsmythe commented on pkg/readiness/ready_tracker_test.go:
```diff
@@ -0,0 +1,238 @@
+	//}
+
+	// Add additional templates/constraints and verify that we remain satisfied
+	// Apply fixtures *before* the controllers are setup.
```
bump, I think this comment may be inaccurate

shomron replied:
Fixed, was copy-pasted from above.

maxsmythe commented on pkg/readiness/object_tracker.go:
```diff
@@ -0,0 +1,258 @@
+
+// Unsatisfied returns all unsatisfied expectations
+func (t *objectTracker) Unsatisfied() []runtime.Object {
+	t.mu.Lock()
```
bump

shomron replied:
Under normal operation I don't expect a lot of contention - each tracker maps to a single controller for writing, plus the readiness probe for reading which won't run that often _under normal circumstances_. I guess since that's an externally facing API it could affect the controllers if someone was slamming `/readyz` in a tight loop. What do you say, still worth going with RWLock?

maxsmythe replied:
IMO we could get away with it, but RLock is strictly better. Not a blocker for me.

shomron replied:
Implemented.

shomron commented:
Sorry about that, GitHub collapsed all those! I'll make another pass.

shomron commented on pkg/readiness/objset.go:
```diff
@@ -0,0 +1,174 @@
+	return o.namespacedName.Name
+}
+
+func (o objKey) SetName(name string) {
```
I'm ripping this out for now. This was a misguided attempt to make a symmetric looking interface for `Unsatisfied() []runtime.Object` and specifically these were additional methods to implement `meta.Object`. The method is not currently used and we should think about the right interface when it is exposed. Also changing `Unsatisfied() -> `unsatisfied()`, or I can remove it altogether for now. Although I'm anticipating a need to introspect which expectations are still unsatisfied through some avenue.

shomron commented:
@maxsmythe whenever you have a chance, I hope I've addressed the remaining concerns. Let me know! 🤞🏻

maxsmythe commented:
LMK what next steps you want to take.

maxsmythe commented:
I saw an email, but I'm unclear if anything changed? LMK if there were any changes and if you want a re-review or to commit.

shomron commented:
@maxsmythe I've changed the comment nits and moved `config.CfgKey` to `keys.Config`. If I'm missing something I apologize, it might be getting lost in the noise.

maxsmythe commented:
ACk, thanks! That should be it. LMK when you want it merged.

shomron commented:
@maxsmythe Woot! Are e2e tests not working? Is the problem unrelated? I could give it a last rebase, otherwise I'm ready for merge from my end.

shomron commented:
Rebased.

maxsmythe commented:
It looks like it's hanging on the e2e tests. Since There is no output, it must be the first test, which is:
"gatekeeper-controller-manager is running" and checks for the ready condition on the gatekeeper-manager pod.

shomron commented:
@maxsmythe I pushed a commit to resolve a race condition while starting up which was blocking the e2e tests. Is there a way for you to re-trigger them? They seem to pass in my fork.

maxsmythe commented:
re-running
